### ISSUE TITLE
ERROR: FATAL ERROR: PCR values failed to match quote's digest!

### ISSUE THREAD

**Issue by lukehinds** [keylime/python-keylime#83]
# Environment
 * OS / version: Fedora 29
 * Processor architecture: x86_64
 * TPM Manufacturer: Infineon SLB9665VQ2.0 TPM v2.0
 * Keylime version: master

# Description 

I tried to run the IMA demo on Fedora 29, but this time doing so on bare metal with a hardware TPM chip.

After uploading the whitelist using the tenant, an exception occurs with `tpm2_deluxequote`

Exception: Command: tpm2_deluxequote -C 0x81000001 -L sha256:15,22+sha1:10 -q 4b6e52706777674b62454841644c46377376345a -m /tmp/tmpa6uTtL -s /tmp/tmpPx7Dq5 -p /tmp/tmpilmRPv -G sha256 -P eGxMQfmwxRkucGVl0czc returned 1, expected 0, output ["ERROR: FATAL ERROR: PCR values failed to match quote's digest!\n", 'ERROR: Error validating calculated PCR composite with quote\n', 'ERROR: Unable to run tpm2_deluxequote\n'

```
Exception: Command: tpm2_deluxequote -C 0x81000001 -L sha256:15,22+sha1:10 -q 4b6e52706777674b62454841644c46377376345a -m /tmp/tmpa6uTtL -s /tmp/tmpPx7Dq5 -p /tmp/tmpilmRPv -G sha256 -P eGxMQfmwxRkucGVl0czc returned 1, expected 0, output ["ERROR: FATAL ERROR: PCR values failed to match quote's digest!\n", 'ERROR: Error validating calculated PCR composite with quote\n', 'ERROR: Unable to run tpm2_deluxequote\n', 'quoted: ff54434780180022000bb321b13de8e32f19e35055ffe7cfb706ac563360b5ce48820976e6941377080f00144b6e52706777674b62454841644c46377376345a000000000001717d0000000000000000010005003f000d190000000002000b030080400004030004000020217d8044568ff67449e1e340b1a413563557be4505e396e9d7586a4f06b0996a\n', 'signature:\n', '  alg: rsassa\n', '  sig: 86016fcc7b8d08f9ab1b048208790f67170e397e1e45ae832c06bdb6894db008f50027fb2fbeaf5eb5de294585d1fbef4b4ad4e4bef3cc4091fcaac52e80d299ec174290eb8d69e9dd541c6afdfe72d59d91adad2538d2c78d1e484e5576111fb54167fb03062b251ee75fa09b28cfaf173489174c7d1fcf71708bdd7ad4f36b519d90da5d31eef46371cc3ac3e8c0ba401391ebcadaa3c93862eb920e24cb444fbaa0105024574468001e6353b7d9df67d959a2ad9e24c4450a98a70c7ebb2b4e09f04ff3bedd037894c7aa02b42b6e1dee0af872ac96257c99146c8e157302fa6bdd7320af1e469349eb3eab02388d436f8c68c3bd70982d9ae396173e875c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\n', 'pcrs:\n', '  sha256:\n', '    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', '    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', '  sha1:\n', '    10: 0x529567C1A8E5D8359711A8C928929D273C4AF811\n', 'calcDigest: fee347c3d458aacd462a8e14da5c91a4ab0a2e044c590fad0b7d25e96ea29710\n']
```

**Comment by lukehinds** [comment]
@jetwhiz @nabilschear , did not create this as bug yet, as not sure if its my TPM. I did run a 'tpm2_takeownership -c', but maybe I need to clear the cmos.

**Comment by jetwhiz** [comment]
Can you uncomment all of the debug output in the `tpm2_util_get_digest_from_quote` function and rebuild/install the tpm2-tools? 

https://github.com/keylime/tpm2-tools/blob/master/lib/tpm2_util.c#L49

**Comment by lukehinds** [comment]
@jetwhiz , sure will do!

**Comment by lukehinds** [comment]
Hopefully this helps:

```
keylime_node                                                                                                                                               
Using config file /etc/keylime.conf
2019-03-15 09:19:09.250 - keylime.secure_mount - INFO - mounting secure storage location /var/lib/keylime/secure on tmpfs
2019-03-15 09:19:09.254 - keylime.tpm2 - INFO - Taking ownership with config provided TPM owner password: keylime
2019-03-15 09:19:09.302 - keylime.tpm2 - INFO - TPM Owner password confirmed: keylime
2019-03-15 09:19:10.534 - keylime.cloudnode - INFO - Node UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-15 09:19:10.613 - keylime.registrar_client - INFO - Node registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-15 09:19:11.055 - keylime.tpm2 - INFO - AIK activated.
2019-03-15 09:19:11.107 - keylime.registrar_client - INFO - Registration activated for node D432FBB3-D2F1-4A97-9EF7-75BD81C00000.
2019-03-15 09:19:11.294 - keylime.cloudnode - INFO - Starting Cloud Node on port 9002 use <Ctrl-C> to stop
2019-03-15 09:19:11.295 - keylime.revocation_notifier - INFO - Waiting for revocation messages on 0mq 127.0.0.1:8992
2019-03-15 09:19:24.259 - keylime.cloudnode - INFO - GET invoked from ('127.0.0.1', 55460) with uri:/quotes/identity?nonce=dpWJtxozmpj9iq06knc4
2019-03-15 09:19:24.259 - keylime.cloudnode - INFO - GET invoked from ('127.0.0.1', 55462) with uri:/quotes/integrity?nonce=ygDqJIIXXpGCCbvB9ovZ&mask=0x408400&vmask=0x808000&partial=0
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 55462)
Traceback (most recent call last):
  File "/usr/lib64/python2.7/SocketServer.py", line 596, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib64/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib64/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/usr/lib64/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib64/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/cloud_node.py", line 126, in do_GET
    quote = tpm.create_quote(nonce, self.server.rsapublickey_exportable, pcrmask, hash_alg)
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tpm2.py", line 724, in create_quote
    retDict = self.__run(command, lock=False, outputpaths=[quotepath.name, sigpath.name, pcrpath.name])
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tpm2.py", line 236, in __run
    raise Exception("Command: %s returned %d, expected %d, output %s"%(cmd, code, expectedcode, retout))
Exception: Command: tpm2_deluxequote -C 0x81000001 -L sha256:15,22+sha1:10 -q 796744714a4949585870474343627642396f765a -m /tmp/tmppO1aZI -s /tmp/tmpAezOLX -p /tmp/tmpx4k0cT -G sha256 -P r3WV76eyl2ykheSwbf4g returned 1, expected 0, output ["ERROR: FATAL ERROR: PCR values failed to match quote's digest!\n", 'ERROR: Error validating calculated PCR composite with quote\n', 'ERROR: Unable to run tpm2_deluxequote\n', 'quoted: ff54434780180022000b33b33b769f6805da1b94787cd02a28662c54de1b3effe7496803c44cee4aa6180014796744714a4949585870474343627642396f765a0000000003a1d8470000000100000000010005003f000d190000000002000b0300804000040300040000202a654d820ddee52057838cb1f8bec9526393b8f2fedd782f9fe6f7b4bd365596\n', 'signature:\n', '  alg: rsassa\n', '  sig: 657936ea841b2d188d3af35597070d4591aed11769531503b48f07ef47449a6ee803c19f2f197797a5d4cb6efc550308626dd96b89c9296e1d440d1d69b37be6fc470b3508299a741608a373fdf3c61f89d8a3234d014ee25136e59af6edba72670031eb3cb881e3fc06452504eaaa5328d0358337983e5da9c73585d2c89b5c66a8ff22014f8fbb6269c1f10114af221ea6f7b28675d2794ac665c5ffb0c777fd852754d1290f4c1ac8cd6e7c62600e7a82b55f514ab018efe986869de28dccd101dbaa56b976afc70e70fcd338d6b29057a60b0d5da48f9987ccb89641a05ebcc43822e38e4eb9f89ce70beace446b43820310416ded4df099cfd1ffb5ff1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\n', 'quoteInfo:\n', '  magic: ff544347\n', '  type: 8018\n', '  nameSize: 34\n', '  extraDataSize: 20\n', '  extraData: 796744714a4949585870474343627642396f765a\n', '  pcrSelCount: 2\n', '    hashAlg: 000b\n', '      sos: 3\n', '    hashAlg: 0004\n', '      sos: 3\n', '  digestSize: 32\n', '  digest: 2a654d820ddee52057838cb1f8bec9526393b8f2fedd782f9fe6f7b4bd365596\n', 'pcrs:\n', '  sha256:\n', '    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', '    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', '  sha1:\n', '    10: 0x2CF4FB5FC5739B40849E90FD0E489120BF811FCE\n', 'calcDigest: f88074e9aab143be02ba62c2a563cb96d529dd3fc8a619070c00a8f4cde7035a\n']
----------------------------------------
2019-03-15 09:19:24.791 - keylime.cloudnode - INFO - GET identity quote returning 200 response.
```


```
keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f /root/exclude.txt --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 --whitelist /root/whitelist.txt --exclude /root/exclude.txt
Using config file /etc/keylime.conf
2019-03-15 09:19:21.274 - keylime.tenant - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-15 09:19:21.274 - keylime.tenant - INFO - Setting up client TLS in /var/lib/keylime/cv_ca
2019-03-15 09:19:21.289 - keylime.tenant - INFO - TPM PCR Mask from policy is 0x408000
2019-03-15 09:19:21.289 - keylime.tenant - INFO - vTPM PCR Mask from policy is 0x808000
2019-03-15 09:19:21.571 - keylime.ima - WARNING - No boot_aggregate value found in whitelist, adding an empty one
2019-03-15 09:19:24.794 - keylime.registrar_client - INFO - Setting up client TLS...
2019-03-15 09:19:24.798 - keylime.registrar_client - WARNING - CAUTION: using default password for private key, please set private_key_pw to a strong password
2019-03-15 09:19:24.829 - keylime.tenant - ERROR - mapping values are not allowed here
  in "<string>", line 23, column 12:
        hashAlg: 000b
               ^
2019-03-15 09:19:25.316 - keylime.tenant - ERROR - Update command response: 404 Unexpected response from Cloud Verifier.
2019-03-15 09:19:25.316 - keylime.tenant - ERROR - Response code 404: instance id not found
```

**Comment by jetwhiz** [comment]
Thanks @lukehinds !   Could you also capture what the output of `tpm2_listpcrs` is that corresponds to this output?  You should be able to run `tpm2_listpcrs` before and/or after running keylime to get a list of all of the PCR values (just in case the values change after running Keylime).

**Comment by lukehinds** [comment]
pre keylime:

```
sha1:
  0 : 0x02E98ABA9AEEE157B3E6C6E5BBFAB3270BF5AD3C
  1 : 0xB4EC92745434AA6717E500B12D6134F00A993799
  2 : 0xF6871775A38E2BC6E9326ECCB4FE50BFDE13B6FC
  3 : 0xB2A83B0EBF2F8374299A5B2BDFC31EA955AD7236
  4 : 0xCAFDD603718D4A31450B0377314B7A76E176EFB1
  5 : 0xEE8798F6888791C620DDCC94E57F5EDFDDD6C638
  6 : 0xB2A83B0EBF2F8374299A5B2BDFC31EA955AD7236
  7 : 0x22C44C57537A3013BE601046910CD91C04B11856
  8 : 0x0F2B31530C4EA039CBF87E042CD9DAABAE48A840
  9 : 0x84FEB6BD7BB08BAC29D15C96B2A43C80D244725A
  10: 0x0FC6AB4C77CA0D07CDD2C421E4C498C17A2930D9
  11: 0x0000000000000000000000000000000000000000
  12: 0x0000000000000000000000000000000000000000
  13: 0x0000000000000000000000000000000000000000
  14: 0x0000000000000000000000000000000000000000
  15: 0x0000000000000000000000000000000000000000
  16: 0x0000000000000000000000000000000000000000
  17: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  18: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  19: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  20: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  21: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  23: 0x0000000000000000000000000000000000000000
sha256:
  0 : 0x7978C7045669E5B7ED3834A70D60139B49DD05F3B6A64799CD6E18BBBF348CD8
  1 : 0x6138A5CD91B91C220C8616FC498DE45CD98E1F8ECC173DD7120BA9AC1120A87B
  2 : 0xF9E7787B241D2333F01E513C3649A7915B0C987AA300B960E26E55A284057923
  3 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
  4 : 0x7A0BD0834A9ECAF392AA4C0E070983D6C863B33B7DA4399702400F0A76F5FCCF
  5 : 0x19AD93ED3C8DE7EC7345ABD664F3016F01E8472ED664AFC0BCD68DD062838A7E
  6 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
  7 : 0x3B7B8385F06EA4D27478E1D60BD7B5797EA4E01A771920A60F19E8DF2204C392
  8 : 0xE636DAAE9F100BDC9A324D6B49BB07190E52E8B95A67432D0F56A20C6994F768
  9 : 0xC9C10FC44E5FBF73B852D0194E991CC8DB78D514FD2E162EEBF01E22240F4814
  10: 0x4B8A84FAA96BBFF56D6DFE3D5BD11FFC9788BA12161482447C953D35EE954DB4
  11: 0x0000000000000000000000000000000000000000000000000000000000000000
  12: 0x0000000000000000000000000000000000000000000000000000000000000000
  13: 0x0000000000000000000000000000000000000000000000000000000000000000
  14: 0x0000000000000000000000000000000000000000000000000000000000000000
  15: 0x0000000000000000000000000000000000000000000000000000000000000000
  16: 0xFD6343DF9D8E831631675B36BBF4039F235F4FC0941EFE2845BADE5D9084E064
  17: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  18: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  19: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  20: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  21: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  23: 0x0000000000000000000000000000000000000000000000000000000000000000
```

run keylime:

```
2019-03-15 19:11:09.974 - keylime.cloudnode - INFO - GET invoked from ('127.0.0.1', 55582) with uri:/quotes/identity?nonce=8ZLJuK7zLeO1TAdyHdEs
2019-03-15 19:11:09.974 - keylime.cloudnode - INFO - GET invoked from ('127.0.0.1', 55584) with uri:/quotes/integrity?nonce=UZVWtpPdpDEB1HmS6GLO&mask=0x408400&vmask=0x808000&partial=0
2019-03-15 19:11:10.233 - keylime.cloudnode - INFO - GET identity quote returning 200 response.
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 55584)
Traceback (most recent call last):
  File "/usr/lib64/python2.7/SocketServer.py", line 596, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib64/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib64/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/usr/lib64/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/lib64/python2.7/BaseHTTPServer.py", line 328, in handle_one_request
    method()
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/cloud_node.py", line 126, in do_GET
    quote = tpm.create_quote(nonce, self.server.rsapublickey_exportable, pcrmask, hash_alg)
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tpm2.py", line 724, in create_quote
    retDict = self.__run(command, lock=False, outputpaths=[quotepath.name, sigpath.name, pcrpath.name])
  File "/usr/lib/python2.7/site-packages/keylime-1.2-py2.7.egg/keylime/tpm2.py", line 236, in __run
    raise Exception("Command: %s returned %d, expected %d, output %s"%(cmd, code, expectedcode, retout))
Exception: Command: tpm2_deluxequote -C 0x81000001 -L sha256:15,22+sha1:10 -q 555a5657747050647044454231486d5336474c4f -m /tmp/tmp1jk3Ii -s /tmp/tmprK78yJ -p /tmp/tmpyTx1OA -G sha256 -P SDWNAicUc8YNSFQoL5Oj returned 1, expected 0, output ["ERROR: FATAL ERROR: PCR values failed to match quote's digest!\n", 'ERROR: Error validating calculated PCR composite with quote\n', 'ERROR: Unable to run tpm2_deluxequote\n', 'quoted: ff54434780180022000bb844a01d5cb16da271e77d1df93e3139420b586fe57099ce759ffdfa6763f56d0014555a5657747050647044454231486d5336474c4f0000000005bd457b0000000100000000010005003f000d190000000002000b0300804000040300040000203de03ee857d60d296cbaa56cb8a9f0d530a95513c12a91e4abbdd0322f137ca5\n', 'signature:\n', '  alg: rsassa\n', '  sig: 0ccf6040fbc5111f41c802bb941e8ada1218aa3d5056179d8080e02348cf9c1401203fc02fe137595f07800a513c5e905bd3c9fda565bb25bfe663ee0d706d1880d343a8a0ae5e90c5394760e75936757f420c07f3b3253b91000478391348b97a750e8b409313b6dbafd2f8a1c144d6bfc5dab9dd58d1dab2d19852503a69637ed9f8bee0eccba8e85be9acf1d5494bf31b97a7171d3a9f21d03cf349687d58afd566ce08244225648d666058e52f5624327c0d2795a1e4416f16d23bd7b72b4c91c637096d6dc0b851a3a590200fa1343aff69a8e09626638ae7ab34f00d2816ef5b0440606fc9b182919cef77851a973245f4457a3340e664009a3b5025b800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\n', 'quoteInfo:\n', '  magic: ff544347\n', '  type: 8018\n', '  nameSize: 34\n', '  extraDataSize: 20\n', '  extraData: 555a5657747050647044454231486d5336474c4f\n', '  pcrSelCount: 2\n', '    hashAlg: 000b\n', '      sos: 3\n', '    hashAlg: 0004\n', '      sos: 3\n', '  digestSize: 32\n', '  digest: 3de03ee857d60d296cbaa56cb8a9f0d530a95513c12a91e4abbdd0322f137ca5\n', 'pcrs:\n', '  sha256:\n', '    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', '    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', '  sha1:\n', '    10: 0x7E4A2CEC74575CB11F95494628B01C3F126080AC\n', 'calcDigest: c64308587da12db4eb3f66a9e5f0628f6720c5f747cb6fd5e1fdf03d25a0c6b5\n']
----------------------------------------
```

pcrlist after:

```
sha1:
  0 : 0x02E98ABA9AEEE157B3E6C6E5BBFAB3270BF5AD3C
  1 : 0xB4EC92745434AA6717E500B12D6134F00A993799
  2 : 0xF6871775A38E2BC6E9326ECCB4FE50BFDE13B6FC
  3 : 0xB2A83B0EBF2F8374299A5B2BDFC31EA955AD7236
  4 : 0xCAFDD603718D4A31450B0377314B7A76E176EFB1
  5 : 0xEE8798F6888791C620DDCC94E57F5EDFDDD6C638
  6 : 0xB2A83B0EBF2F8374299A5B2BDFC31EA955AD7236
  7 : 0x22C44C57537A3013BE601046910CD91C04B11856
  8 : 0x0F2B31530C4EA039CBF87E042CD9DAABAE48A840
  9 : 0x84FEB6BD7BB08BAC29D15C96B2A43C80D244725A
  10: 0xDAF721ED477F38F108DA6D1EE9C321FE1BCDBAD0
  11: 0x0000000000000000000000000000000000000000
  12: 0x0000000000000000000000000000000000000000
  13: 0x0000000000000000000000000000000000000000
  14: 0x0000000000000000000000000000000000000000
  15: 0x0000000000000000000000000000000000000000
  16: 0x0000000000000000000000000000000000000000
  17: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  18: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  19: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  20: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  21: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  23: 0x0000000000000000000000000000000000000000
sha256:
  0 : 0x7978C7045669E5B7ED3834A70D60139B49DD05F3B6A64799CD6E18BBBF348CD8
  1 : 0x6138A5CD91B91C220C8616FC498DE45CD98E1F8ECC173DD7120BA9AC1120A87B
  2 : 0xF9E7787B241D2333F01E513C3649A7915B0C987AA300B960E26E55A284057923
  3 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
  4 : 0x7A0BD0834A9ECAF392AA4C0E070983D6C863B33B7DA4399702400F0A76F5FCCF
  5 : 0x19AD93ED3C8DE7EC7345ABD664F3016F01E8472ED664AFC0BCD68DD062838A7E
  6 : 0x3D458CFE55CC03EA1F443F1562BEEC8DF51C75E14A9FCF9A7234A13F198E7969
  7 : 0x3B7B8385F06EA4D27478E1D60BD7B5797EA4E01A771920A60F19E8DF2204C392
  8 : 0xE636DAAE9F100BDC9A324D6B49BB07190E52E8B95A67432D0F56A20C6994F768
  9 : 0xC9C10FC44E5FBF73B852D0194E991CC8DB78D514FD2E162EEBF01E22240F4814
  10: 0xBE1B1CBB01E5F97321D58C0AE7DDA7268BBED60F44B856EB56E29A3266AF0C91
  11: 0x0000000000000000000000000000000000000000000000000000000000000000
  12: 0x0000000000000000000000000000000000000000000000000000000000000000
  13: 0x0000000000000000000000000000000000000000000000000000000000000000
  14: 0x0000000000000000000000000000000000000000000000000000000000000000
  15: 0x0000000000000000000000000000000000000000000000000000000000000000
  16: 0xFD6343DF9D8E831631675B36BBF4039F235F4FC0941EFE2845BADE5D9084E064
  17: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  18: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  19: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  20: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  21: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  23: 0x0000000000000000000000000000000000000000000000000000000000000000
```

Hope that helps!

**Comment by jetwhiz** [comment]
It looks like your PCR10/IMA measurements might be changing very quickly, which is causing a race condition where deluxequote asks the TPM for a quote and then asks the TPM for its PCR values.  When it tries to compare the PCR values to what is contained in the quote (to verify they are correct), they differ since PCR10 is being updated so much and so quickly. 

What kind of workload is running while you're doing this?  Is something else on your machine writing to PCR10 other than IMA?  How much time passed between running pcrlist, keylime and pcrlist?

**Comment by lukehinds** [comment]
> What kind of workload is running while you're doing this? Is something else on your machine writing to PCR10 other than IMA? How much time passed between running pcrlist, keylime and pcrlist?

Keylime is the only workload and that is IMA monitoring. Nothing else is using the TPM to my knowledge. I could try to get specific times, but as I ran pcrlist manually (not using any script) and started keylime in between, it it would be approximately 30-40 seconds (run tpm2_pcrlist, run keylime_verifier/registe/node, run tpm2_pcrlist).

**Comment by jetwhiz** [comment]
The thing that will make debugging this tricky is that we're not getting consistent PCR values, so I'm not sure if the issue is: 
1. Your TPM is calculating the aggregate PCR differently than the others we've tested (in a way that I can't yet figure out), or 
2. PCR10 changes its value in-between quote and pcrlist being called in `tpm2_deluxequote`

The way deluxequote works is that it makes two calls to the TPM:  one for the quote (which contains a signed aggregate of the PCR values we asked to be included in the quote) and one for the PCR values themselves (so we can send them to Keylime to check individually).  There is a possibility for a (what should be rare) race condition, where the PCR values might change between these two calls.  

If you notice in your exception line, you get: 
````bash
Exception: Command: tpm2_deluxequote 
 ... 
'  digestSize: 32\n', 
'  digest: 3de03ee857d60d296cbaa56cb8a9f0d530a95513c12a91e4abbdd0322f137ca5\n', 
'pcrs:\n', 
'  sha256:\n', 
'    15: 0x0000000000000000000000000000000000000000000000000000000000000000\n', 
'    22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n', 
'  sha1:\n', 
'    10: 0x7E4A2CEC74575CB11F95494628B01C3F126080AC\n', 
'calcDigest: c64308587da12db4eb3f66a9e5f0628f6720c5f747cb6fd5e1fdf03d25a0c6b5\n']
````

Note that the PCR10 (sha1) value there (`0x7E4A2CEC74575CB11F95494628B01C3F126080AC`) doesn't match either of your PCR10 values that you got when running `tpm2_pcrlist` manually (`0x0FC6AB4C77CA0D07CDD2C421E4C498C17A2930D9` and `0xDAF721ED477F38F108DA6D1EE9C321FE1BCDBAD0`).  This means that I don't know which PCR values your TPM is using to compute its aggregate hash for sure. 

But by manually calculating the aggregate hash for those three PCR values, `SHA256(0x0000000000000000000000000000000000000000000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF7E4A2CEC74575CB11F95494628B01C3F126080AC)`, I get the value that tpm2_deluxequote expected (`calcDigest:  c64308587da12db4eb3f66a9e5f0628f6720c5f747cb6fd5e1fdf03d25a0c6b5`).  

For some reason, this differs from the value provided by your TPM in the quote (`digest:  3de03ee857d60d296cbaa56cb8a9f0d530a95513c12a91e4abbdd0322f137ca5`), which is why this failure is being returned.  

The only things I can think of is that your PCR10 is changing so quickly that it triggers this rare race condition, or your TPM is calculating the aggregate hash in a non-standard way (which I can't confirm, since I don't know for sure which PCR values it is using to calculate its digest). 

Can you try running `tpm2_pcrlist` immediately before bringing the node up and immediately after so that the PCR10 values before, during and after are all consistent?  If PCR10 changes very quickly then it might take a few attempts.  

Also, can you check your IMA `ascii_runtime_measurements` file to see if something is being rapidly extended into the TPM by IMA?  Maybe something is misconfigured there?

**Comment by lukehinds** [comment]
Good news, we can close this. It was resolved by upgrading to kernel 5.0.5-200.fc29.x86_64 which has the following patch:

https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg1959715.html
